File: src/java/hep/lcio/example/LCTools.java
/*
 * LCTools.java
 *
 * Created on May 4, 2003, 9:48 PM
 */
package hep.lcio.example;

import hep.lcio.event.LCFloatVec;
import hep.lcio.event.LCIntVec;
import hep.lcio.event.LCStrVec;

import hep.lcio.event.CalorimeterHit;
import hep.lcio.event.LCCollection;
import hep.lcio.event.LCEvent;
import hep.lcio.event.LCIO;
import hep.lcio.event.MCParticle;
import hep.lcio.event.SimCalorimeterHit;
import hep.lcio.event.SimTrackerHit;
import hep.lcio.event.TPCHit;


/**
 *
 * @author  Tony Johnson
 */
public class LCTools
{
   public static void dumpEvent(LCEvent evt)
   {
      // the event:
      System.out.println("event  : " + evt.getEventNumber() + " - run " + evt.getRunNumber() + " detector : " + evt.getDetectorName() + " - collections  : ");

      String[] strVec = evt.getCollectionNames();

      // loop over collections:
      for (int j = 0; j < strVec.length; j++)
      {
         String name = strVec[j];

         System.out.print("     " + name + " " + evt.getCollection(name).getTypeName() + " : ");

         LCCollection col = evt.getCollection(name);

         // print SimCalorimeterHit collections:
         if (evt.getCollection(name).getTypeName().equals(LCIO.SIMCALORIMETERHIT))
         {
            int nHits = col.getNumberOfElements();
            System.out.print(nHits + " hits - first hit: ");

            int nPrint = (nHits > 0) ? 1 : 0;

            if (nPrint == 0)
               System.out.println();
            for (int i = 0; i < nPrint; i++)
            {
               SimCalorimeterHit hit = (SimCalorimeterHit) col.getElementAt(i);

               float[] x = hit.getPosition();

               System.out.print("    hit -  e: " + hit.getEnergy() + "  pos: " + x[0] + ", " + x[1] + ", " + x[2]);

               if (hit.getNMCParticles() > 0)
                  System.out.print(" - mc: " + hit.getParticleCont(0).getPDG());

               System.out.println();
            }
         }

         // print CalorimeterHit collections:
         else if (evt.getCollection(name).getTypeName().equals(LCIO.CALORIMETERHIT))
         {
            int nHits = col.getNumberOfElements();
            System.out.print(nHits + " hits - first hit: ");

            int nPrint = (nHits > 0) ? 1 : 0;

            if (nPrint == 0)
               System.out.println();
            for (int i = 0; i < nPrint; i++)
            {
               CalorimeterHit hit = (CalorimeterHit) col.getElementAt(i);

               float[] x = hit.getPosition();

               System.out.print("    hit -  e: " + hit.getEnergy() + "  pos: " + x[0] + ", " + x[1] + ", " + x[2]);
               System.out.println();
            }
         }
         else if (evt.getCollection(name).getTypeName().equals(LCIO.SIMTRACKERHIT))
         {
            int nHits = col.getNumberOfElements();
            System.out.print(nHits + " hits : ");

            int nPrint = (nHits > 0) ? 1 : 0;

            if (nPrint == 0)
               System.out.println();
            for (int i = 0; i < nPrint; i++)
            {
               SimTrackerHit hit = (SimTrackerHit) col.getElementAt(i);

               double[] x = hit.getPosition();

               int pdg = -999;
               MCParticle part = hit.getMCParticle();
               if( part != null)
                 pdg = part.getPDG();

               System.out.print("    hit -  dEdx: " + hit.getdEdx() + "  mc: " + pdg + "  pos: " + x[0] + ", " + x[1] + ", " + x[2]);
            }
            System.out.println();
         }
         else if (evt.getCollection(name).getTypeName().equals(LCIO.TPCHIT))
         {
         // check whether we are dealing with raw data
            int flag = col.getFlag();
            int mask = 1<<LCIO.TPCBIT_RAW;
            boolean raw = (flag & mask) == mask;

             
            int nHits = col.getNumberOfElements();
            System.out.print(nHits + " hits : ");

            int nPrint = (nHits > 0) ? 1 : 0;

            if (nPrint == 0)
               System.out.println();
            for (int i = 0; i < nPrint; i++)
            {
               TPCHit hit = (TPCHit) col.getElementAt(i);
	       if(raw){
		   int nRawWords = hit.getNRawDataWords();
		   System.out.print( nRawWords +" raw data words:");
		   for(int rawWord = 0; rawWord< nRawWords; ++rawWord) {
		       System.out.print( " | " + hit.getRawDataWord(rawWord) );
		   }
		   System.out.println(" |");
	       }
            else
            {
               System.out.print("    hit -  id: " + hit.getCellID() + " time: " + hit.getTime()
                + "  charge: " + hit.getCharge() + ", quality: " + hit.getQuality() );
                }
            }
             
         }
         else if (evt.getCollection(name).getTypeName().equals(LCIO.LCFLOATVEC))
         {
            int nHits = col.getNumberOfElements();
            System.out.print(nHits + " vectors: ");

            int nPrint = (nHits > 0) ? 1 : 0;

            if (nPrint == 0)
               System.out.println();
            for (int i = 0; i < nPrint; i++)
            {
               LCFloatVec vec = (LCFloatVec) col.getElementAt(i);

               System.out.print(" values(" + i + "): ");

               float[] data = vec.toArray();
               for (int k = 0; k < data.length; k++)
                  System.out.print(data[k] + ", ");
               System.out.println();
            }
         }
         else if (evt.getCollection(name).getTypeName().equals(LCIO.LCINTVEC))
         {
            int nHits = col.getNumberOfElements();
            System.out.print(nHits + " vectors: ");

            int nPrint = (nHits > 0) ? 1 : 0;

            if (nPrint == 0)
               System.out.println();
            for (int i = 0; i < nPrint; i++)
            {
               LCIntVec vec = (LCIntVec) col.getElementAt(i);

               System.out.print(" values(" + i + "): ");

               int[] data = vec.toArray();
               for (int k = 0; k < data.length; k++)
                  System.out.print(data[k] + ", ");
               System.out.println();
            }            
         }         
         else if (evt.getCollection(name).getTypeName().equals(LCIO.LCSTRVEC))
         {
            int nHits = col.getNumberOfElements();
            System.out.print(nHits + " vectors: ");

            int nPrint = (nHits > 0) ? 1 : 0;

            if (nPrint == 0)
               System.out.println();
            for (int i = 0; i < nPrint; i++)
            {
               LCStrVec vec = (LCStrVec) col.getElementAt(i);

               System.out.print(" values(" + i + "): ");

               String[] data = vec.toArray();
               for (int k = 0; k < data.length; k++)
                  System.out.print(data[k] + ", ");
               System.out.println();
            }            
         }         
         else if (evt.getCollection(name).getTypeName().equals(LCIO.MCPARTICLE))
         {
            int nHits = col.getNumberOfElements();
            System.out.print(nHits + " particles : ");

            int nPrint = (nHits > 0) ? 1 : 0;

            if (nPrint == 0)
               System.out.println();
            for (int i = 0; i < nPrint; i++)
            {
               MCParticle part = (MCParticle) col.getElementAt(i);

               System.out.println("           " + part.getPDG() + " p: " + part.getMomentum()[0] + ", " + part.getMomentum()[1] + ", " + part.getMomentum()[2]);
            }
         }
         else
         {
            System.out.println();
         }
      }
      System.out.println();
   }
}

File: src/cpp/src/IMPL/LCTOOLS.cc
#include "IMPL/LCTOOLS.h"
#include "EVENT/LCCollection.h"
#include "EVENT/SimCalorimeterHit.h"
#include "EVENT/SimTrackerHit.h"
#include "EVENT/LCIO.h"
#include "EVENT/MCParticle.h"
#include "DATA/LCFloatVec.h"

#include "IMPL/LCFlagImpl.h"

using namespace std ;
using namespace EVENT ;
using namespace DATA ;


namespace IMPL {

  static int MAX_HITS = 1000 ;


  void LCTOOLS::dumpEventDetailed(const LCEvent* evt){
    
    // the event:
    cout << endl 
	 << "==============================================================" << endl ;
    cout << "        Event  : " << evt->getEventNumber() 
	 << " - run:  "         << evt->getRunNumber()
         << " - timestamp "     << evt->getTimeStamp()  << endl ;
    cout << "==============================================================" << endl ;    

    cout << " detector : "      << evt->getDetectorName() << endl ;
    

    const std::vector< std::string >* strVec = evt->getCollectionNames() ;

    // loop over all collections:
    std::vector< std::string >::const_iterator name ;
    
    for( name = strVec->begin() ; name != strVec->end() ; name++){
    
      const LCCollection* col = evt->getCollection( *name ) ;
    
      cout << endl 
	   << " collection name : " << *name 
	   << endl ;

      // call the detailed print functions depending on type name
      if( evt->getCollection( *name )->getTypeName() == LCIO::MCPARTICLE ){
	  
	printMCParticles( col ) ;

      }
      else if( evt->getCollection( *name )->getTypeName() == LCIO::SIMTRACKERHIT ){
	  
	printSimTrackerHits( col ) ;

      }
      else if( evt->getCollection( *name )->getTypeName() == LCIO::SIMCALORIMETERHIT ){
	  
	printSimCalorimeterHits( col ) ;

      }


    }

  }

  void LCTOOLS::dumpEvent(const LCEvent* evt){
    
    // the event:
    cout << "event  : " 
	 << evt->getEventNumber() 
	 << " - run " << evt->getRunNumber()
	 << " detector : "  <<  evt->getDetectorName()
	 << " - collections  : "
	 << endl ;
    
    const std::vector< std::string >* strVec = evt->getCollectionNames() ;

    // loop over collections:
    for( std::vector< std::string >::const_iterator name = strVec->begin() ; name != strVec->end() ; name++){
    
      cout << "     " <<  *name <<   " " 
	   <<   evt->getCollection( *name )->getTypeName() << " : "  ; 
    
      const LCCollection* col = evt->getCollection( *name ) ;
    
    
      // print SimCalorimeterHit collections:
      if(evt->getCollection( *name )->getTypeName() == LCIO::SIMCALORIMETERHIT ){
      
	int nHits =  col->getNumberOfElements() ;
	cout << nHits << " hits - first hit: " ;
	int nPrint = nHits>0 ? 1 : 0 ;
	int flag = col->getFlag() ;

	if(!nPrint ) cout << endl ;
	for( int i=0 ; i< nPrint ; i++ ){
	
	  const SimCalorimeterHit* hit = 
	    dynamic_cast<const SimCalorimeterHit*>( col->getElementAt( i ) ) ;
	
	  cout << "    hit -  e: "  << hit->getEnergy() ;

	  if( LCFlagImpl(flag).bitSet( LCIO::CHBIT_LONG ) ){

	    const float* x =  hit->getPosition() ;
	    cout << "  pos: " << x[0] << ", " << x[1] << ", " << x[2] << endl ;   

	  } else{
	    cout << "  short hits: position not available! " << hex << flag << dec << endl ;
	  }
	}
      
	// print the MCParticle collection
      } else if(evt->getCollection( *name )->getTypeName() == LCIO::SIMTRACKERHIT ){
      
	int nHits =  col->getNumberOfElements() ;
	cout << nHits << " hits : " ;
	int nPrint = nHits>0 ? 1 : 0 ;

	if(!nPrint ) cout << endl ;
	for( int i=0 ; i< nPrint ; i++ ){
	  const SimTrackerHit* hit = 
	    dynamic_cast<const SimTrackerHit*>( col->getElementAt( i ) ) ;
	
	  const double* x =  hit->getPosition() ;
	  cout << "    hit -  dEdx: " 
	       << hit->getdEdx() 
	       << "  mc: " << hit->getMCParticle()->getPDG() 
	       << "  pos: " 
	       << x[0] << ", " << x[1] << ", " << x[2] 
	       << endl ;   

	}
      } else if(evt->getCollection( *name )->getTypeName() == LCIO::LCFLOATVEC ){
	
	int nHits =  col->getNumberOfElements() ;
	cout << nHits << " vectors: " ;
	int nPrint = nHits>0 ? 1 : 0 ;
	
	if(!nPrint ) cout << endl ;
	for( int i=0 ; i< nPrint ; i++ ){
	  const LCFloatVec* vec = 
	    dynamic_cast<const LCFloatVec*>( col->getElementAt( i ) ) ;
	  
	  cout << " values(" << i << "): " ;
	  for(unsigned int k=0 ; k< vec->size() ; k++ )
	    cout <<  (*vec)[k]  << ", " ;
	  cout << endl ;   

	}
      } else if(evt->getCollection( *name )->getTypeName() == LCIO::MCPARTICLE ){
      
      
	int nHits =  col->getNumberOfElements() ;
	cout << nHits << " particles : " ;
	int nPrint = nHits>0 ? 1 : 0 ;

	if(!nPrint ) cout << endl ;
	for( int i=0 ; i< nPrint ; i++ ){
	  const MCParticle* part = 
	    dynamic_cast<const MCParticle*>( col->getElementAt( i ) ) ;
	
	  cout << "           " << part->getPDG() << " p: " 
	       <<  part->getMomentum()[0]  << ", "
	       <<  part->getMomentum()[1]  << ", "
	       <<  part->getMomentum()[2]  
	       << endl ;	
	
	}

      } else {
	cout << endl ;
      } 
    }
    cout << endl ;
  }


  void LCTOOLS::printSimTrackerHits(const EVENT::LCCollection* col ){

  }

  void LCTOOLS::printSimCalorimeterHits(const EVENT::LCCollection* col ){

  }

  void LCTOOLS::printMCParticles(const EVENT::LCCollection* col ) {

    if( col->getTypeName() != LCIO::MCPARTICLE ){

      cout << " collection not of type " << LCIO::MCPARTICLE << endl ;
      return ;
    }

    cout << endl 
	 << "--------------- " << "print out of "  << LCIO::MCPARTICLE << " collection "
	 << "--------------- " << endl ;

    cout << endl 
	 << "  flag:  0x" << hex  << col->getFlag() << dec << endl ;
    
    int nParticles =  col->getNumberOfElements() ;


    // get a list of all mother particles
    std::vector<const MCParticle*> moms ; 
    for( int k=0; k<nParticles; k++){

      // get the particle from the collection - needs a cast !
      const MCParticle* part =  dynamic_cast<const MCParticle*>( col->getElementAt( k ) ) ;
      
      if( part->getParentData() == 0 )
	moms.push_back( part ) ;
    }

    // now loop over mothers and print daughters recursively
    int index = 0 ;
    std::vector<const MCParticle*>::const_iterator  mom ; 
    for( mom = moms.begin() ; mom != moms.end() ; mom++){

      cout << index++ << ": "
	   <<  (*mom)->getPDG() << " | " 
	   <<  (*mom)->getMomentum()[0]  << ", "
	   <<  (*mom)->getMomentum()[1]  << ", "
	   <<  (*mom)->getMomentum()[2]  
	   << endl ;	

      //      index  = printDaughterParticles( *mom, index ) ;
    }

    int nPrint = nParticles > MAX_HITS ? MAX_HITS : nParticles ;

    if(!nPrint ) cout << endl ;
    for( int i=0 ; i< nPrint ; i++ ){

      // get the particle from the collection - needs a cast !
      const MCParticle* part =  dynamic_cast<const MCParticle*>( col->getElementAt( i ) ) ;
      


      cout <<  part->getPDG() << " | " 
	   <<  part->getMomentum()[0]  << ", "
	   <<  part->getMomentum()[1]  << ", "
	   <<  part->getMomentum()[2]   << "   "
	   << "me: " <<  part 
	   << " mom: " << part->getParentData() ;
      
      if( part->getNumberOfDaughters() > 0 ) 
	cout <<  " - daughter: " << part->getDaughter(0) ;
      cout    << endl ;	
      
    }

    cout << endl 
	 << "-------------------------------------------------------------------------------- " 
	 << endl ;
  }

  int LCTOOLS::printDaughterParticles(const MCParticle* part, int index){

    // print this particles daughters
    for(int i=0; i<part->getNumberOfDaughters();i++){
      
      cout << index++ << ": "
	   <<  part->getPDG() << " | " 
	   <<  part->getMomentum()[0]  << ", "
	   <<  part->getMomentum()[1]  << ", "
	   <<  part->getMomentum()[2]  
	   << endl ;	
    }
    // and now the same for the grand daughters
    for(int i=0; i<part->getNumberOfDaughters();i++){

      index = printDaughterParticles( part->getDaughter(i) , index ) ;
    }

    return index ;

  }

}; // namespace

